File: ekodide/vizinhanca.py
"""A vizinhança do Ekodide: descobre quem está na rede SEM digitar IP.

Cada ponta que está escutando (`ekodide serve`) "grita" de tempos em tempos um
pacotinho UDP no broadcast da LAN: *"oi, sou o <nome>, atendo na porta <porta>"*.
Quem quer enviar só ESCUTA por alguns segundos e monta a lista de aparelhos.

O IP NÃO viaja no pacote — quem escuta lê o endereço do remetente (`recvfrom`).
Assim, se o IP mudar (DHCP), a descoberta continua certa: o nome resolve pro IP
de agora, não pro de ontem.

Continua zero-dependência (só `socket`/`json`/`threading`/`time`) e determinístico:
isto só anuncia presença e escuta — quem decide enviar é quem aciona. NÃO carrega
o segredo nem conteúdo; é só a plaquinha de "estou aqui".
"""
from __future__ import annotations

import json
import socket
import threading
import time

# Porta separada da transferência (8778): aqui só trafega o anúncio de presença.
PORTA_DESCOBERTA = 8779
# Marca pra ignorar pacote de outro programa que caia na mesma porta.
MARCA = "ekodide-vizinho-1"
# Pra onde o anúncio é jogado. 255.255.255.255 = broadcast do enlace local (a LAN).
BROADCAST = "255.255.255.255"


def _pacote(nome: str, porta: int) -> bytes:
    return json.dumps({"marca": MARCA, "nome": nome, "porta": porta}).encode("utf-8")


def anunciar(
    nome: str,
    porta: int,
    parar: threading.Event | None = None,
    intervalo: float = 2.0,
    enderecos: list[str] | None = None,
) -> threading.Event:
    """Fica gritando 'estou aqui' até `parar` ser acionado. Bloqueia — pra rodar
    junto do recebedor use `anunciar_em_thread`. Devolve o Event de parada.

    Por padrão grita no broadcast da LAN (pros outros aparelhos) E no 127.0.0.1
    (pra você conferir na mesma máquina — broadcast não faz loopback pra si mesmo)."""
    parar = parar or threading.Event()
    enderecos = enderecos if enderecos is not None else [BROADCAST, "127.0.0.1"]
    msg = _pacote(nome, porta)
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    try:
        while not parar.is_set():
            for endereco in enderecos:
                try:
                    s.sendto(msg, (endereco, PORTA_DESCOBERTA))
                except OSError:
                    pass  # rede caiu/sem enlace agora — tenta de novo no próximo ciclo
            parar.wait(intervalo)
    finally:
        s.close()
    return parar


def anunciar_em_thread(nome: str, porta: int, **kw) -> threading.Event:
    """Sobe o anúncio numa thread daemon e volta na hora. Acione o Event devolvido
    (`.set()`) pra parar de anunciar."""
    parar = threading.Event()
    alvo = lambda: anunciar(nome, porta, parar=parar, **kw)
    threading.Thread(target=alvo, daemon=True).start()
    return parar


def procurar(timeout: float = 2.5, porta: int = PORTA_DESCOBERTA) -> list[dict]:
    """Escuta os anúncios por `timeout` segundos e devolve os aparelhos vistos:
    lista de {"nome", "ip", "porta"}, ordenada por nome. Um aparelho só conta uma
    vez (o anúncio mais recente vence, caso a porta mude)."""
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    try:
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)  # vários ouvintes no mesmo host
    except (AttributeError, OSError):
        pass
    s.bind(("", porta))
    s.settimeout(0.4)
    achados: dict[str, dict] = {}
    fim = time.monotonic() + timeout
    try:
        while time.monotonic() < fim:
            try:
                dados, remetente = s.recvfrom(2048)
            except socket.timeout:
                continue
            except OSError:
                break
            try:
                p = json.loads(dados)
            except ValueError:
                continue
            if not isinstance(p, dict) or p.get("marca") != MARCA:
                continue
            nome = p.get("nome")
            porta_aparelho = p.get("porta")
            if not isinstance(nome, str) or not isinstance(porta_aparelho, int):
                continue
            achados[nome] = {"nome": nome, "ip": remetente[0], "porta": porta_aparelho}
    finally:
        s.close()
    return sorted(achados.values(), key=lambda d: d["nome"])


def url_de(aparelho: dict) -> str:
    """Monta a URL de transferência ('http://IP:PORTA') a partir de um achado."""
    return f"http://{aparelho['ip']}:{aparelho['porta']}"

File: android/app/src/main/kotlin/com/ekodide/android/net/Vizinhanca.kt
package com.ekodide.android.net

import com.ekodide.android.core.CanonicalJson
import com.ekodide.android.core.JsonParser
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.SocketTimeoutException
import kotlin.concurrent.thread

/**
 * A vizinhança: descobre quem está na rede SEM digitar IP. Espelho do vizinhanca.py.
 *
 * O celular (recebedor passivo) fica "gritando" de tempos em tempos um pacotinho UDP no
 * broadcast da LAN: *"oi, sou o <nome>, atendo na porta <porta>"*. O PC só ESCUTA por
 * alguns segundos e monta a lista. O IP NÃO viaja no pacote — quem escuta lê o endereço
 * do remetente; assim, se o IP mudar (DHCP), o nome resolve pro IP de agora.
 *
 * O pacote é só JSON parseado (sem HMAC, sem segredo, sem conteúdo) — a plaquinha de
 * "estou aqui". Por isso não precisa ser byte-idêntico ao Python: basta JSON válido com
 * marca/nome/porta, que os dois lados leem com o JsonParser.
 *
 * NOTA Android: ENVIAR (anunciar) não exige nada especial; RECEBER broadcast no Wi-Fi
 * pede MulticastLock — isso entra com o foreground service (Etapa 4).
 */
object Vizinhanca {
    // Porta separada da transferência (8778): aqui só trafega o anúncio de presença.
    const val PORTA_DESCOBERTA = 8779
    // Marca pra ignorar pacote de outro programa que caia na mesma porta.
    const val MARCA = "ekodide-vizinho-1"
    // 255.255.255.255 = broadcast do enlace local (a LAN).
    const val BROADCAST = "255.255.255.255"

    data class Vizinho(val nome: String, val ip: String, val porta: Int)

    /** Sinal de parada (espelha o threading.Event do Python): para de anunciar com `parar()`. */
    class Parada {
        @Volatile private var ativo = true
        private val trava = Object()
        val parado: Boolean get() = !ativo

        fun parar() {
            synchronized(trava) { ativo = false; trava.notifyAll() }
        }

        /** Dorme até `ms` OU até alguém chamar `parar()` (o que vier primeiro). */
        fun aguardar(ms: Long) {
            synchronized(trava) { if (ativo) trava.wait(ms) }
        }
    }

    private fun pacote(nome: String, porta: Int): ByteArray =
        CanonicalJson.encode(mapOf("marca" to MARCA, "nome" to nome, "porta" to porta))
            .toByteArray(Charsets.UTF_8)

    /**
     * Fica gritando "estou aqui" até `parada` ser acionada. BLOQUEIA — pra rodar junto do
     * servidor use `anunciarEmThread`. Por padrão grita no broadcast da LAN (pros outros)
     * E no 127.0.0.1 (pra conferir na mesma máquina — broadcast não faz loopback pra si).
     */
    fun anunciar(
        nome: String,
        porta: Int,
        parada: Parada = Parada(),
        intervaloMs: Long = 2000L,
        enderecos: List<String> = listOf(BROADCAST, "127.0.0.1"),
        portaDestino: Int = PORTA_DESCOBERTA,
    ): Parada {
        val msg = pacote(nome, porta)
        val s = DatagramSocket()
        s.broadcast = true
        try {
            while (!parada.parado) {
                for (endereco in enderecos) {
                    try {
                        val pkt = DatagramPacket(
                            msg, msg.size, InetAddress.getByName(endereco), portaDestino,
                        )
                        s.send(pkt)
                    } catch (_: Exception) {
                        // rede caiu/sem enlace agora — tenta de novo no próximo ciclo
                    }
                }
                parada.aguardar(intervaloMs)
            }
        } finally {
            s.close()
        }
        return parada
    }

    /** Sobe o anúncio numa thread daemon e volta na hora. Pare com `.parar()` no retorno. */
    fun anunciarEmThread(
        nome: String,
        porta: Int,
        intervaloMs: Long = 2000L,
        enderecos: List<String> = listOf(BROADCAST, "127.0.0.1"),
        portaDestino: Int = PORTA_DESCOBERTA,
    ): Parada {
        val parada = Parada()
        thread(isDaemon = true, name = "ekodide-vizinhanca") {
            anunciar(nome, porta, parada, intervaloMs, enderecos, portaDestino)
        }
        return parada
    }

    /**
     * Escuta os anúncios por `timeoutMs` e devolve os aparelhos vistos: {nome, ip, porta},
     * ordenado por nome. Um aparelho só conta uma vez (o anúncio mais recente vence).
     */
    fun procurar(timeoutMs: Long = 2500L, porta: Int = PORTA_DESCOBERTA): List<Vizinho> {
        val s = DatagramSocket(null)
        s.reuseAddress = true
        s.bind(InetSocketAddress(porta))
        s.soTimeout = 400
        val achados = LinkedHashMap<String, Vizinho>()
        val fim = System.nanoTime() + timeoutMs * 1_000_000
        val buf = ByteArray(2048)
        try {
            while (System.nanoTime() < fim) {
                val pkt = DatagramPacket(buf, buf.size)
                try {
                    s.receive(pkt)
                } catch (_: SocketTimeoutException) {
                    continue
                } catch (_: Exception) {
                    break
                }
                val texto = String(pkt.data, pkt.offset, pkt.length, Charsets.UTF_8)
                val p = try { JsonParser.parse(texto) } catch (_: Exception) { continue }
                if (p !is Map<*, *> || p["marca"] != MARCA) continue
                val nome = p["nome"] as? String ?: continue
                val portaAp = (p["porta"] as? Long)?.toInt() ?: continue
                val ip = pkt.address?.hostAddress ?: continue
                achados[nome] = Vizinho(nome, ip, portaAp)
            }
        } finally {
            s.close()
        }
        return achados.values.sortedBy { it.nome }
    }

    /** Monta a URL de transferência ('http://IP:PORTA') a partir de um achado. */
    fun urlDe(v: Vizinho): String = "http://${v.ip}:${v.porta}"
}
